unicon_backend/evaluator/tasks/programming/templates/mpi_sandbox.py

```diff
@@ -1,14 +1,41 @@
+import hashlib
 import importlib
 import importlib.util as importlib_util
 import io
 import multiprocessing
+import os
 from contextlib import redirect_stderr, redirect_stdout
 from multiprocessing import Process as MPProcess
 from multiprocessing import Queue as MPQueue
 from typing import Final
 
 TASK_Q_STOP_SIGNAL: Final[str] = "STOP"
 
+DIR_PATH = os.path.dirname(os.path.realpath(__file__))
+
+
+class FileIntegrityError(ValueError):
+    pass
+
+
+def verify_file_md5(file_path: str, expected_hash: str) -> None:
+    abs_file_path = os.path.join(DIR_PATH, file_path)
+    if not os.path.exists(abs_file_path):
+        raise FileIntegrityError(f"File integrity check failed (file not found): {file_path}")
+    md5_hash = hashlib.md5()
+    with open(abs_file_path, "rb") as file:
+        # Read the file in chunks to handle large files efficiently
+        for chunk in iter(lambda: file.read(4096), b""):
+            md5_hash.update(chunk)
+
+    calculated_hash = md5_hash.hexdigest()
+    if calculated_hash.lower() != expected_hash.lower():
+        raise FileIntegrityError(
+            f"File integrity check failed (mismatch hash)\n"
+            f"> Expected: {expected_hash.lower()}\n"
+            f"> Got: {calculated_hash.lower()}"
+        )
```

Comment:
how are we catching this error? will the runner be responsible and will there be a new evaluation status for integrity errors or will it be grouped under `RTE`?

Reply:
At the moment, this is likely being caught under `RTE` because there’s no specific handling for it in the runner or backend. Ideally, it should be classified as an integrity error. However, I’m not sure how to implement that without parsing the error message from stderr. Any thoughts?

Reply:
we could edit the [global exception hook](https://github.com/uniconhq/backend/blob/main/unicon_backend/evaluator/tasks/programming/templates/global_except_hook.py) to catch these new errors and return a custom exit code. the runners will then subsequently handle this new exit code and interpret it as a integrity error

Reply:
Okay, will do that then.